File: apps/vscode/src/tools/track/styling/setTrackColor.ts
/**
 * Set Track Color tool implementation.
 * Sets the display color for track features.
 */

import type { TrackFeature } from '@debrief/schemas';
import { isTrackFeature } from '@debrief/schemas';
import type { MCPToolDefinition } from '../../../types/tool';

export interface SetTrackColorParams {
  color: string;
}

export const toolDefinition: MCPToolDefinition = {
  name: 'set-track-color',
  description: 'Sets the display color for track features.',
  inputSchema: {
    type: 'object',
    properties: {
      features: { type: 'array', items: { type: 'object' } },
      params: { type: 'object', properties: { color: { type: 'string', description: 'CSS color value', 'x-debrief-param-type': 'NamedColor' } } },
    },
  },
  annotations: {
    'debrief:selectionRequirements': [{ kind: 'TRACK', min: 1 }],
    'debrief:category': 'track/styling',
    'debrief:version': '1.0.0',
    'debrief:outputKind': 'mutation/track/styled',
    'debrief:uiCategory': 'style',
  },
};

export function execute(
  features: TrackFeature[],
  params: SetTrackColorParams,
): TrackFeature[] {
  const { color } = params;
  if (!color) {
    throw new Error('color parameter is required');
  }

  const modified: TrackFeature[] = [];

  for (const feature of features) {
    if (!isTrackFeature(feature)) {
      continue;
    }

    // Defensively handle missing style — real features from disk may lack it
    const style = feature.properties.style ?? {};
    const line = style.line ?? { stroke: true, color: '#3388ff', weight: 3, opacity: 1.0 };
    line.color = color;
    style.line = line;
    feature.properties.style = style;

    modified.push(feature);
  }

  if (modified.length === 0) {
    throw new Error('No track features found in input');
  }

  return modified;
}

File: services/calc/debrief_calc/tools/track/styling/set_track_color.py
"""Set Track Color tool."""

from __future__ import annotations

from typing import Any

from debrief_calc.models import (
    ContextType,
    GeoJSONFeatureDict,
    SelectionContext,
    ToolCategoryEnum,
    ToolParameter,
)
from debrief_calc.registry import tool


@tool(
    name="set-track-color",
    description="Sets the display color for track features. Modifies the line color property of each track's style.",
    input_kinds=["TRACK"],
    output_kind="mutation/track/styled",
    context_type=ContextType.MULTI,
    category=ToolCategoryEnum.style,
    parameters=[
        ToolParameter(
            name="color",
            type="string",
            description="CSS color value to apply",
            required=True,
            param_type="NamedColor",
        ),
    ],
)
def set_track_color(context: SelectionContext, params: dict[str, Any]) -> list[GeoJSONFeatureDict]:
    """
    Set the display color for track features.

    Args:
        context: SelectionContext with one or more track features
        params: Parameters dict with 'color' (required)

    Returns:
        List of modified track features with updated line color
    """
    color = params.get("color")
    if not color:
        raise ValueError("color parameter is required")

    modified = []
    for feature in context.features:
        props = feature.get("properties", {})
        if props.get("kind") != "TRACK":
            continue

        # Initialize style hierarchy with defaults
        style = props.setdefault("style", {})
        line = style.setdefault(
            "line",
            {
                "stroke": True,
                "color": "#3388ff",
                "weight": 3,
                "opacity": 1.0,
            },
        )
        line["color"] = color

        modified.append(feature)

    if not modified:
        raise ValueError("No track features found in input")

    return modified
